### DISCUSSION TITLE
anon role, RLS policies, status code, ownership

### DISCUSSION THREAD

**Discussion by ethicnology** [supabase/supabase#1606] [Questions] [answered]
I'm discovering supabase and as a student i would like to understand deeper the mindset of supabase covering these subjects :

* anon role
* RLS
* insert with specific value

## anon role
* I don't get the point of the **anon key/role** why would we allow data fetching to an anon user ? (except during test)
* Shouldn't, default be restrictive as possible ? (privacy by design)
* Can we create a client directly with user credential in order to get the JWT ? (bypass anon key)

### table profiles 
| id | email |
|----|---------|
| 48aa59cc-7742-4e64-adc7-cefa78fe14bd | a@a.fr |
| 4b78ec2a-a83f-4ac3-b411-78a726c1953b | b@b.fr |

On each auth.users insertion it will insert uid and email in public.profiles [using this.](https://supabase.io/docs/guides/auth#create-a-publicusers-table)

### table personas
This table contains all the personas which an user can play
| id | profile_id | model | x | y | z |
|----|--------------|----------|---|---|---|
| 25 | 48aa59cc-7742-4e64-adc7-cefa78fe14bd | a | 0 | 0 | 70 |
| 36 | 4b78ec2a-a83f-4ac3-b411-78a726c1953b | a | 0 | 0 | 70 |  

### table personas_details
| id | persona_id | something |
|----|-----------------|---------------|
| 1 | 25 | idk |

### policies
* **profiles**  : ``` auth.uid() = id ALL PERMISSIVE```
*  **personas** : ``` auth.uid() = profile_id ALL PERMISSIVE```

### issues
Because of PostgREST, i was thinking that status code will be available, but when i perform these queries i have no results and no errors.
#### 401 expected
I expect an error if the **anon user** attempt to fetch the table.
I tried to add role policy ``` auth.role() = 'authenticated' ALL PERMISSIVE/ RESTRICTIVE  ``` with no success.
```ts
  const { data, error } = await SUPABASE.from('profiles').select('*')
  console.log(data, error) // [] null
```

#### 404 expected
I expect an error if **authenticated user** try to fetch something which is not found
```ts
  const { data, error } = await SUPABASE.from('profiles').select('*').eq('email', 'not@in.table')
  console.log(data, error) // [] null
```

#### ownership in deeper relations
For tables **profiles and personas** it's easy to check ownership thanks to ``` auth.uid() ``` and the presence of this **uid in profiles primary key and personas foreign key**.
How should i do for the example table **personas_details** which has ``` persona_id as foreign key ``` ?
I was thinking of a policy which embed a SELECT but i don't know if i can catch the **personas_details.id** value in the policy  
```sql
1 = (SELECT COUNT(*) FROM public.personas WHERE profile_id = auth.uid() AND id = '36') /* how to catch id value ?*/
```

#### ensure foreign key insertion
During an insertion on table **personas_details** i would like to put the good **foreign_key persona_id** without rely on the front-end. I want to check ownership the foreign key value 
In custom REST API we can insert the good foreign key based on user JWT.
In PostgreSQL I've found many methodologies, what is the **best practice ?** : 
```sql
SELECT id FROM public.personas WHERE id = ? /* requested persona_id for insertion */
```
* INSTEAD OF INSERT
* BEFORE INSERT
* AFTER INSERT 
* [Middleware ?](https://supabase.io/docs/guides/auth#you-dont-have-to-use-policies) (is there a tutorial for supabase middleware integration ?)

**Comment by steve-chavez** [comment, marked as answer]
> anon role
> Shouldn't, default be restrictive as possible ? (privacy by design)

Yes, this is a bug that will be fixed. `anon` shouldn't have any privilege by default. See this previous discussion https://github.com/supabase/supabase/discussions/1493#discussioncomment-724842 for how to revoke anon privileges.

> I expect an error if the anon user attempt to fetch the table.

Since `anon` has the privilege to SELECT, right now there's no error. But if you apply the fix above there will be.

> 404 expected
> I expect an error if authenticated user try to fetch something which is not found

If you use the [single()](https://supabase.io/docs/reference/javascript/single) modifier then you'll get an error in this case.

> ensure foreign key insertion
> During an insertion on table personas_details i would like to put the good foreign_key persona_id without rely on the front-end.

Creating a SQL function and calling it through [rpc()](https://supabase.io/docs/reference/javascript/rpc) would be the way now. Later on, it will be possible to do this through the client libraries(once https://github.com/PostgREST/postgrest/issues/818 is solved).

**Reply by ethicnology** [reply]
Thanks a lot for all these advises.
Should i disable INSERT on the concerned table each time i choose to implement a custom insertion through rpc ? 
Do you have an idea to check ownership on this table with policies or should i replace SELECT by a stored procedure too ?

**Reply by steve-chavez** [reply]
> Should i disable INSERT on the concerned table each time i choose to implement a custom insertion through rpc ?

Yeah, that looks apt for ensuring all clients inserts only pass through your sql function.

> Do you have an idea to check ownership on this table with policies or should i replace SELECT by a stored procedure too ?
> ownership in deeper relations

No, that looks it should be possible with regular RLS policies. I haven't fully understood your problem there, but have a look at the [policies examples](https://supabase.io/docs/guides/auth#policies-with-joins) we have in the docs.

Let me know if that helps.

**Reply by ethicnology** [reply]
Finally, for the foreign key i've made this :
```sql
-- trigger the function on insert into personas_details 
create trigger on_personas_details_insert
  before insert on public.personas_details
  for each row execute procedure public.insert_personas_details();

-- verify new.persona_id foreign key belongs to the user (policy auth.uid() = personas.profile_id)   
create or replace function public.insert_personas_details()
returns trigger as $$
begin
  select personas.id into strict new.persona_id from public.personas where id=new.persona_id;
  return new;
end;
$$ language plpgsql;
```

Thanks to policy ``` auth.uid() = profile_id ALL PERMISSIVE ``` on personas, select will only return persona owned by the user.
[The **STRICT** option ensure the select must return exactly one row or throw an error.](https://www.postgresql.org/docs/13/plpgsql-statements.html)
